### PR TITLE
Frontend combine, tensor ops

### PR DESCRIPTION
**Pull request by Icemole**
Creates `combine()`, handles tensor operations (e.g. Tensor + Tensor), and some more improvements.

### CODE REVIEW

File: returnn/tensor/tensor.py [hunk 1]
```diff
@@ -21,11 +21,13 @@
...
 
 from __future__ import annotations
-from typing import Optional, Sequence, Tuple, Generic, TypeVar
+from typing import Optional, Sequence, Tuple, Generic, TypeVar, Union
+from numpy import number as numpy_number
 
 from .dim import Dim
 from ._tensor_extra import _TensorExtra, _TensorMixin
 from . import _tensor_extra
+from .. import frontend_api as _frontend_api
```
Comment (albertz) [inline returnn/tensor/tensor.py:30]: As I told you, remove this here, or make it conditional with TYPE_CHECKING.

File: returnn/tensor/tensor.py [hunk 2]
```diff
@@ -123,5 +125,141 @@ def raw_tensor(self, value: Optional[RawTensorType]):
         self._raw_tensor = value
         self.sanity_check(assume_complete=False)
 
+    def _tensor_op_sanity_check(self, other: Union[_frontend_api.RawTensorTypes, Tensor]):
```
Comment (albertz) [inline returnn/tensor/tensor.py:128]: Remove this function here. This should be directly called in `combine` and `compare`.

File: returnn/tensor/tensor.py [hunk 3]
```diff
@@ -123,5 +125,141 @@ def raw_tensor(self, value: Optional[RawTensorType]):
...
+
+    def __rtruediv__(self, other: Union[_frontend_api.RawTensorTypes, Tensor]) -> Tensor:
+        self._tensor_op_sanity_check(other)
+        if isinstance(other, (int, float, numpy_number)):
+            if other == 1:
+                return self
+            else:
+                assert not self.dtype.startswith(
+                    "int"
+                ), "Dividing a Tensor of type int by an integer is disallowed. Please convert the Tensor to float."
```
Comment (albertz) [inline returnn/tensor/tensor.py:215]: Remove such logic here. This logic should be in `combine`.

File: returnn/tensor/tensor.py [hunk 4]
```diff
@@ -123,5 +125,141 @@ def raw_tensor(self, value: Optional[RawTensorType]):
...
+
+    def __truediv__(self, other: Union[_frontend_api.RawTensorTypes, Tensor]) -> Tensor:
+        self._tensor_op_sanity_check(other)
+        if isinstance(other, (int, float, numpy_number)):
+            if other == 1:
+                return self
+            else:
+                assert not self.dtype.startswith(
+                    "int"
+                ), "Dividing a Tensor of type int by an integer is disallowed. Please convert the Tensor to float."
```
Comment (albertz) [inline returnn/tensor/tensor.py:203]: Remove this here. This should be in `combine`.

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -261,6 +261,7 @@ jobs:
           - TEST=demos RETURNN_DISABLE_TF=1
           - TEST=PTDataset
           - TEST=torch_frontend
+          - TEST=torch_internal_frontend
```
Comment (albertz) [inline .github/workflows/main.yml:264]: As discussed, the tests should be torch independent. We should not have torch specific tests here.

**Comment by albertz** [comment]
All the merge commits should be removed. You rather should rebase.

**Comment by albertz** [comment]
Please check tests. You should configure your environment correctly. It seems you do not have Black configured. Please refer to the wiki.

File: tests/test_torch_internal_frontend.py [hunk 6]
```diff
@@ -4,15 +4,16 @@
...
 
 from returnn.tensor import Tensor, Dim
 from returnn.torch.frontend import TorchFrontend as rf
 
 
 def test_compare_eq():
-    a_raw = rf.Tensor([2.0, 2.0, 2.0])
-    b_raw = rf.Tensor([1.0, 2.0, 3.0])
+    a_raw = torch.tensor([2.0, 2.0, 2.0])
+    b_raw = torch.tensor([1.0, 2.0, 3.0])
```
Comment (albertz) [inline tests/test_torch_internal_frontend.py:16]: Don't do this. Use only `rf` functions.

**Comment by albertz** [comment]
Why did you add those functions to `Tensor` and not to `_TensorMixin`?

**Comment by Icemole** [comment]
> Why did you add those functions to `Tensor` and not to `_TensorMixin`?

Because `compare` and `combine` required a `Tensor` as parameter, which is not always the case if I do `_TensorMixin.__add__` as a `_TensorMixin` is not always a `Tensor`.
But if you want I can change those to `_TensorMixin` and place the `__add__` functions etc in `_TensorMixin`.

File: .github/workflows/main.yml (hunk 5 above)
Comment (Icemole) [reply .github/workflows/main.yml:264]: Since we need this PR merged I won't do this here, we can implement the unified tests later on.

**Comment by albertz** [comment]
> > Why did you add those functions to `Tensor` and not to `_TensorMixin`?
> 
> Because `compare` and `combine` required a `Tensor` as parameter, which is not always the case if I do `_TensorMixin.__add__` as a `_TensorMixin` is not always a `Tensor`. But if you want I can change those to `_TensorMixin` and place the `__add__` functions etc in `_TensorMixin`.

`_TensorMixin` is supposed to be a mixin, i.e. `self` is always a `Tensor`. You can also specify this explicitly via type hints. See some of the other methods.

File: tests/test_torch_internal_frontend.py (hunk 6 above)
Comment (Icemole) [reply tests/test_torch_internal_frontend.py:16]: I will use `convert_to_tensor`.

File: .github/workflows/main.yml (hunk 5 above)
Comment (albertz) [reply .github/workflows/main.yml:264]: Ok.

File: tests/test_torch_internal_frontend.py (hunk 6 above)
Comment (Icemole) [reply tests/test_torch_internal_frontend.py:16]: For these I believe we should have `rf.constant()` as we discussed in the past, which is not available yet. If this is correct, I suggest we temporarily leave it like this.

Comment (albertz) [reply tests/test_torch_internal_frontend.py:19]: Just add `rf.constant` then. Or use `rf.range` or sth else. This really should not be complicated.

File: returnn/frontend_api.py [hunk 7]
```diff
@@ -42,6 +42,20 @@ def convert_to_tensor(value: Union[Tensor, T, RawTensorTypes]) -> Tensor[T]:
         """
         raise NotImplementedError
 
+    @staticmethod
+    def _tensor_op_sanity_check(t1: Union[RawTensorTypes, Tensor], t2: Union[RawTensorTypes, Tensor]):
+        """
+        Performs basic checks common to all _TensorMixin binary operations,
+        like _TensorMixin.__eq__ or _TensorMixin.__add__.
```
Comment (albertz) [inline returnn/frontend_api.py:49]: There should be an empty line after the descriptions, before the params. (Keep consistency.)

File: returnn/frontend_api.py [hunk 8]
```diff
@@ -42,6 +42,20 @@ def convert_to_tensor(value: Union[Tensor, T, RawTensorTypes]) -> Tensor[T]:
...
+        :param t1:
+        :param t2:
+        """
+        if isinstance(t1, _t.Tensor) and isinstance(t2, _t.Tensor):
```
Comment (albertz) [inline returnn/frontend_api.py:53]: The way you call `_tensor_op_sanity_check`, you always expect that you have `isinstance(t1, _t.Tensor) and isinstance(t2, _t.Tensor)`, right? Then don't make this an `if` check, but rather an `assert`.

File: returnn/tensor/_tensor_extra.py [hunk 9]
```diff
@@ -609,6 +610,116 @@ def __repr__(self):
...
+            return self
+        return self.raw_frontend.combine(self, "*", other)
+
+    def __rmul__(self: Tensor, other: Union[_frontend_api.RawTensorTypes, Tensor]) -> Tensor:
+        if isinstance(other, (int, float, numpy_number)) and other == 1:
+            return self
+        return self.raw_frontend.combine(other, "*", self)
+
+    def __truediv__(self: Tensor, other: Union[_frontend_api.RawTensorTypes, Tensor]) -> Tensor:
+        if isinstance(other, (int, float, numpy_number)):
```
Comment (albertz) [inline returnn/tensor/_tensor_extra.py:667]: This check is duplicated here, right? You already do the same check in `combine`. Then just leave it in `combine`, and remove it here?

File: returnn/tensor/_tensor_extra.py [hunk 10]
```diff
@@ -609,6 +610,116 @@ def __repr__(self):
...
+                return self
+            else:
+                assert not self.dtype.startswith(
+                    "int"
+                ), "Dividing a Tensor of type int by an integer is disallowed. Please convert the Tensor to float."
+
+        return self.raw_frontend.combine(self, "/", other)
+
+    def __rtruediv__(self: Tensor, other: Union[_frontend_api.RawTensorTypes, Tensor]) -> Tensor:
+        if isinstance(other, (int, float, numpy_number)):
```
Comment (albertz) [inline returnn/tensor/_tensor_extra.py:678]: Same.

File: returnn/frontend_api.py [hunk 11]
```diff
@@ -85,6 +100,67 @@ def compare(
...
+            Not all the dims of a and b need to be specified here, and there could also be other dims in the dim_order.
+        :return: element-wise combination of a and b
+        """
+        a = cls.convert_to_tensor(a)
+        b = cls.convert_to_tensor(b)
+        cls._tensor_op_sanity_check(a, b)
+        # Truediv checks for now
+        if kind in {"truediv", "/"}:
+            if isinstance(a, (int, float, numpy.number)) and a == 1:
+                return b
```
Comment (albertz) [inline returnn/frontend_api.py:130]: This is wrong. 1/b != b.

File: returnn/frontend_api.py [hunk 12]
```diff
@@ -85,6 +100,67 @@ def compare(
...
+        :param allow_broadcast_all_sources: if True, it is allowed that neither a nor b has all dims of the result.
+            Not needed when out_dims is specified explicitly.
+        :param dim_order: defines the order of the resulting dims. if None, it is automatically inferred from a and b.
+            Not all the dims of a and b need to be specified here, and there could also be other dims in the dim_order.
+        :return: element-wise combination of a and b
+        """
+        a = cls.convert_to_tensor(a)
+        b = cls.convert_to_tensor(b)
+        cls._tensor_op_sanity_check(a, b)
+        # Truediv checks for now
```
Comment (albertz) [inline returnn/frontend_api.py:127]: It should be consistent. You should move all such checks here.

File: returnn/tensor/_tensor_extra.py [hunk 13]
```diff
@@ -5,6 +5,7 @@
 
 from __future__ import annotations
 from typing import TYPE_CHECKING, Optional, Union, Type, Sequence, Tuple, List, Dict
+from numpy import number as numpy_number
```
Comment (albertz) [inline returnn/tensor/_tensor_extra.py:8]: This is very uncommon. Just do `import numpy` instead, and then refer to it as `numpy.number`.

**Comment by albertz** [comment]
> All the merge commits should be removed. You rather should rebase.

Can you still do that? And also clean up some unnecessary commits (reorder, squash).

File: returnn/frontend_api.py [hunk 14]
```diff
@@ -85,6 +100,67 @@ def compare(
...
+        if kind in {"truediv", "/"}:
+            if isinstance(a, (int, float, numpy.number)) and a == 1:
```
[the remark this answers is not in the document]
Comment (Icemole) [reply returnn/frontend_api.py:129]: In fact when we convert the scalar value (int/float/np.number) to a `Tensor`, we don't have to do all these checks, do we?

[the remark this answers is not in the document]
Comment (Icemole) [reply returnn/frontend_api.py:129]: The `truediv` checks are correct, I mean the rest of the checks that return `self`: multiplication * 1, addition + 0, etc.

[the remark this answers is not in the document]
Comment (albertz) [reply returnn/frontend_api.py:129]: I don't understand. It's just the same?

[the remark this answers is not in the document]
Comment (Icemole) [reply returnn/frontend_api.py:129]: I hadn't seen your latest message. Locally I've moved this check before creating tensors `a` and `b`, which should be correct. What I meant was that I found redundant to return `self` with the identity operation.

[the remark this answers is not in the document]
Comment (albertz) [reply returnn/frontend_api.py]: I don't understand. What exactly is redundant?

**Comment by Icemole** [comment]
I've squashed my commits into 5 relevant commits. In these I've also applied the feedback you suggested.

**Pull request merged**